### PR TITLE
Nn 2862 cells with capacity api

### PR DESCRIPTION
**Pull request by Nimphal**
Get cells with capacity for agency and group them by location grouping configuration.

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/controllers/LocationController.kt [hunk 1]
```diff
@@ -27,4 +33,17 @@ class LocationController(private val locationService: LocationService) {
                        @ApiParam(value = "The group name", required = true) @PathVariable("name") name: String): List<Location>
     = locationService.getCellLocationsForGroup(agencyId, name)
 
+  @GetMapping("/cellsWithCapacity/{agencyId}/{group}")
+  @ApiOperation(value = "List of cells by group at agency location which have capacity.", notes = "List of cells  by group at agency location which have capacity.", nickname = "getCellsWithCapacityForGroup")
+  @ApiResponses(value = [
+    ApiResponse(code = 200, message = "OK", response = CellWithAttributes::class, responseContainer = "List"),
+    ApiResponse(code = 400, message = "Invalid request.", response = ErrorResponse::class, responseContainer = "List"),
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/controllers/LocationController.kt:40]: I don't think you need the responseContainer for the errors. Might be worth checking.

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/CellAttribute.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.justice.digital.hmpps.whereabouts.model
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import io.swagger.annotations.ApiModel
+import io.swagger.annotations.ApiModelProperty
+import javax.validation.constraints.NotBlank
+import javax.validation.constraints.NotNull
+
+/**
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/CellAttribute.kt:9]: Has this been left in by mistake?

Comment (Nimphal) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/model/CellAttribute.kt:7]: Copy paste, copy paste... Fixed :)

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/controllers/LocationController.kt (hunk 1 above)
Comment (Nimphal) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/controllers/LocationController.kt]: Looks like you don't, have cleaned up this controller.

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/LocationService.kt [hunk 3]
```diff
@@ -17,8 +18,24 @@ class LocationService(
           .map { it.copy(description = it.description.formatLocation()) }
           .toList()
 
+  fun getCellsWithCapacityForGroup(agencyId: String, groupName: String, attribute: String?): List<CellWithAttributes> =
+          elite2ApiService.getCellsWithCapacity(agencyId, attribute)
+                  .filter{ (locationGroupService.locationGroupFilter(agencyId, groupName)::test)(it.mapToLocation(agencyId)) }
+                  .toMutableList()
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/LocationService.kt:24]: Is this undone by the toList()?

Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/LocationService.kt:24]: Actually, do we need to make this mutable at all?

**Review by JamieDraperUK — DISMISSED** [review]
LGTM

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/LocationService.kt (hunk 3 above)
Comment (Nimphal) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/LocationService.kt]: This is why I love code reviews. No, we don't need to be making into two different kinds of lists :D Have removed.

**Review by JamieDraperUK — APPROVED** [review]
LGTM

**Pull request merged**